File: test/timestamps.test.ts
import { mkdtempSync, readFileSync, writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { fileURLToPath } from "node:url";
import { describe, expect, it } from "vitest";
import { loadCached, storeCached } from "../src/pricing/cache.ts";

// Cache fetchedAt boundary conformance, driven by the SHARED fixtures in
// golden/pricing/timestamps/. The stamp is one wire contract read and written
// by seven facades; TS and the native core must agree on every boundary.

type ReadCase = { name: string; stamp: string; epochMs: number | null };
type Cases = { read: ReadCase[]; write: number[] };

const casesPath = fileURLToPath(
  new URL("../golden/pricing/timestamps/cases.json", import.meta.url),
);
const cases = JSON.parse(readFileSync(casesPath, "utf8")) as Cases;

function writeStamp(dir: string, name: string, stamp: string): void {
  writeFileSync(join(dir, name), JSON.stringify({ fetchedAt: stamp, payload: {} }));
}

describe("cache timestamp boundary read (shared gold)", () => {
  for (const kase of cases.read) {
    it(kase.name, () => {
      const dir = mkdtempSync(join(tmpdir(), "skopli-ts-ts-"));
      const name = "pricing-openrouter.json";
      writeStamp(dir, name, kase.stamp);
      if (kase.epochMs === null) {
        expect(loadCached(dir, name, 0, 0)).toBeNull();
        return;
      }
      // At now == epochMs the age is zero, not stale; one ms later it is stale.
      // The pair pins the parsed epoch to the millisecond.
      expect(loadCached(dir, name, 0, kase.epochMs)?.stale).toBe(false);
      expect(loadCached(dir, name, 0, kase.epochMs + 1)?.stale).toBe(true);
    });
  }
});

describe("cache timestamp boundary write (shared gold)", () => {
  for (const ms of cases.write) {
    it(`writes integral-ms for ${ms}`, () => {
      const dir = mkdtempSync(join(tmpdir(), "skopli-ts-tsw-"));
      const stamp = storeCached(dir, "pricing-openrouter.json", {}, ms);
      expect(stamp).toHaveLength(24);
      expect(stamp.endsWith("Z")).toBe(true);
      expect(Date.parse(stamp)).toBe(ms);
    });
  }
});

File: sdks/go/skopli/timestamps_test.go
// Cache fetchedAt boundary conformance, driven by the SHARED fixtures in
// golden/pricing/timestamps/. The stamp is one wire contract read and written
// by seven facades; a stamp is read through this facade's own cache path
// (loadCached / storeCached) so Go and the native core agree on every boundary.
package skopli

import (
	"encoding/json"
	"os"
	"path/filepath"
	"testing"
	"time"
)

type tsReadCase struct {
	Name    string `json:"name"`
	Stamp   string `json:"stamp"`
	EpochMs *int64 `json:"epochMs"`
}

type tsCases struct {
	Read  []tsReadCase `json:"read"`
	Write []int64      `json:"write"`
}

func loadTimestampCases(t *testing.T) tsCases {
	t.Helper()
	p := filepath.Join(repoRoot(t), "golden", "pricing", "timestamps", "cases.json")
	b, err := os.ReadFile(p)
	if err != nil {
		t.Fatalf("read timestamps cases: %v", err)
	}
	var cases tsCases
	if err := json.Unmarshal(b, &cases); err != nil {
		t.Fatalf("decode timestamps cases: %v", err)
	}
	return cases
}

func writeStamp(t *testing.T, dir, name, stamp string) {
	t.Helper()
	body, err := json.Marshal(cacheFile{FetchedAt: stamp, Payload: json.RawMessage(`{}`)})
	if err != nil {
		t.Fatalf("marshal cache file: %v", err)
	}
	if err := os.WriteFile(filepath.Join(dir, name), body, 0o644); err != nil {
		t.Fatalf("write cache file: %v", err)
	}
}

func TestTimestampReadConformance(t *testing.T) {
	const name = "pricing-openrouter.json"
	for _, kase := range loadTimestampCases(t).Read {
		t.Run(kase.Name, func(t *testing.T) {
			dir := t.TempDir()
			writeStamp(t, dir, name, kase.Stamp)
			if kase.EpochMs == nil {
				if got := loadCached(dir, name, 0, time.UnixMilli(0)); got != nil {
					t.Fatalf("invalid stamp parsed to %+v, want nil", got)
				}
				return
			}
			// A 10 ms guard band brackets the parsed instant: gross misparses
			// (an ignored offset, a rejected pre-epoch) flip these assertions,
			// while sub-millisecond representation differences never do.
			ms := *kase.EpochMs
			if got := loadCached(dir, name, 0, time.UnixMilli(ms-10)); got == nil || got.stale {
				t.Fatalf("at now-10ms: got %+v, want fresh", got)
			}
			if got := loadCached(dir, name, 0, time.UnixMilli(ms+10)); got == nil || !got.stale {
				t.Fatalf("at now+10ms: got %+v, want stale", got)
			}
		})
	}
}

func TestTimestampWriteConformance(t *testing.T) {
	const name = "pricing-openrouter.json"
	for _, ms := range loadTimestampCases(t).Write {
		t.Run(time.UnixMilli(ms).UTC().Format(time.RFC3339Nano), func(t *testing.T) {
			dir := t.TempDir()
			stamp, err := storeCached(dir, name, []byte(`{}`), time.UnixMilli(ms))
			if err != nil {
				t.Fatalf("storeCached: %v", err)
			}
			if len(stamp) != 24 {
				t.Fatalf("stamp %q length = %d, want 24", stamp, len(stamp))
			}
			if stamp[len(stamp)-1] != 'Z' {
				t.Fatalf("stamp %q not Z-suffixed", stamp)
			}
			if got := loadCached(dir, name, 0, time.UnixMilli(ms-10)); got == nil || got.stale {
				t.Fatalf("at now-10ms: got %+v, want fresh", got)
			}
			if got := loadCached(dir, name, 0, time.UnixMilli(ms+10)); got == nil || !got.stale {
				t.Fatalf("at now+10ms: got %+v, want stale", got)
			}
		})
	}
}
